Homework 3

JeRusakow opened the pull request.

JeRusakow committed:
Task1 solved

JeRusakow committed:
Configuration files introdused to git repo

JeRusakow committed:
Task2 solved

JeRusakow committed:
Task3 solved

JeRusakow committed:
Task4 solved

iamthen0ise commented on .isort.cfg:
```diff
@@ -0,0 +1,4 @@
+[settings]
+known_third_party =
```
(optional) можно удалить) все равно ничего нет

JeRusakow replied:
Isort не разрешает полностью избавиться от этой строчки. Сам добавляет её, хоть и пустую

iamthen0ise commented on homework3/task01/dec_cache.py:
```diff
@@ -0,0 +1,17 @@
+            if args not in result_dict.keys() or result_dict[args][1] == 0:
+                result_dict[args] = [func_orig(*args), times]
+
+            result_dict[args][1] -= 1
```
если будет возможность как-то, например с DefaultDict, описать что является ключем и значением словаря, я буду рад) а то не зная текста задачи можно голову сломать)

JeRusakow replied:
Использовать словарь со значением по умолчанию не получится, поскольку сам словарь объявляется на уровень выше, где args ещё неизвестны. 
Для понятности не придумал ничего лучше, чем написать docstring'и и разделить это сложное условие на два более простых и снабдить их комментариями

iamthen0ise commented on homework3/task02/voodoo_calc.py:
```diff
@@ -0,0 +1,40 @@
+            shared_val.value += res
+
+    processes = [
+        mp.Process(target=worker, args=(shared_value, val)) for val in range(501)
```
нет желания как-то ограничить кол-во воркеров? не у всех на компьютере есть 500 ядер)

JeRusakow replied:
Желание-то есть, да) Однако не понимаю, что можно для этого ещё использовать
Дело в том, что slow_calculations обязательно засыпает на несколько секунд, которые хорошо бы потратить на что-то полезное. Поначалу пытался использовать Pool.map для 501 процесса, но он просто резервирует ядра под них, где они выполняются всё равно последовательно, и спящие процессы не вытесняются.  
А при создании 501 независимого процесса бремя распоряжения процессорным временем ложится на ОС, с чем она справляется

JeRusakow replied:
Вернулся к идее с Pool.map(). При создании пула устанавливаю число одновременно выполняющихся процессов в 50 (по умолчанию, вынес это число в параметры функции), а дальше все дела так же отдаются на откуп планировщику ОС

iamthen0ise commented:
На зачет все готово) предлагаю добить вопросы по чистоте кода и с мультипроком на более крутую оценку.

JeRusakow committed:
1. Added docstrings for decorator and wrapper-fuction
2. Complex conditional operator divided into two simplier ones in order to make the code clearer

JeRusakow committed:
Optimized process spawning

The functions now spawns only 50 (by default or more if stated) processes simultaniously to perform the task.
This quantitys is enough to match the time requirement (60 sec).

JeRusakow committed:
Minor changes to meet the code convention

JeRusakow committed:
Isort makes changes to its own config

It doesn't allow to completely detele "known_third_party=" line
Don't know why :\

iamthen0ise commented:
В данном случае оптимизация только все ухудшила — теперь данные не считаются в параллели, а медленный код просто запускается 4 раза в разных процессах). Нужно вернуть парраллельность + очень советую посмотреть на тактику когда большой массив разбивают на части и считают каждую отдельно. иначе это ничуть не лучше чем запускать все в главном потоке.

JeRusakow committed:
Multiprocessing task fix

JeRusakow committed:
Doc update

JeRusakow committed:
Code cleaned

JeRusakow committed:
Worker function moved a level up

JeRusakow committed:
Merge branch 'main' into hw3
